cms/src/kemri.rs

```diff
@@ -0,0 +1,65 @@
+//! KEMRecipientInfo-related types
+
+use crate::{
+    content_info::CmsVersion,
+    enveloped_data::{EncryptedKey, RecipientIdentifier, UserKeyingMaterial},
+};
+use const_oid::ObjectIdentifier;
+use der::{asn1::OctetString, Sequence};
+use spki::AlgorithmIdentifierOwned;
+
+/// From [RFC9629 Section 3]
+/// ```text
+///   id-ori OBJECT IDENTIFIER ::= { iso(1) member-body(2) us(840)
+///     rsadsi(113549) pkcs(1) pkcs-9(9) smime(16) 13 }
+///
+///   id-ori-kem OBJECT IDENTIFIER ::= { id-ori 3 }
+/// ```
+/// [RFC9629 Section 3]: https://datatracker.ietf.org/doc/html/rfc9629#section-3
+pub const ID_ORI_KEM: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113549.1.9.16.13.3");
+
+/// The `KEMRecipientInfo` type is defined in [RFC9629 Section 3]
+/// ```text
+///   KEMRecipientInfo ::= SEQUENCE {
+///     version CMSVersion,  -- always set to 0
+///     rid RecipientIdentifier,
+///     kem KEMAlgorithmIdentifier,
+///     kemct OCTET STRING,
+///     kdf KeyDerivationAlgorithmIdentifier,
+///     kekLength INTEGER (1..65535),
+///     ukm [0] EXPLICIT UserKeyingMaterial OPTIONAL,
+///     wrap KeyEncryptionAlgorithmIdentifier,
+///     encryptedKey EncryptedKey }
+/// ```
+/// [RFC9629 Section 3]: https://datatracker.ietf.org/doc/html/rfc9629#section-3
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct KemRecipientInfo {
+    pub version: CmsVersion,
+    pub rid: RecipientIdentifier,
+    pub kem: AlgorithmIdentifierOwned,
+    pub kem_ct: OctetString,
+    pub kdf: AlgorithmIdentifierOwned,
+    pub kek_length: u16,
```

Comment:
any opinion about making that a `NonZero<u16>`?

That will require the `DecodeValue`/`EncodeValue` for that (and a bump release of der) but from a type system standpoint this is better?

I don't know if that's worth the effort.

Reply:
Checking for non-zero would only satisfy part of the requirement (which I'd tend to leave to the decryption component). Here's the snip from the RFC:

```
Implementations MUST confirm that the value provided is consistent with the key-encryption algorithm identified in the wrap field below.
```

Reply:
That would look something like: https://github.com/carl-wallace/formats/compare/kemri...baloo:formats:baloo/kemri